src/Grpc.AspNetCore.Server/Internal/CallHandlers/ServerCallHandlerBase.cs

```diff
@@ -46,10 +48,115 @@ public Task HandleCallAsync(HttpContext httpContext)
+
+                    if (handleCallTask.IsCompletedSuccessfully)
+                    {
+                        return serverCallContext.EndCallAsync();
+                    }
+                    else
+                    {
+                        return AwaitHandleCall(serverCallContext, Method, handleCallTask);
+                    }
+                }
+                else
+                {
+                    return HandleCallWithDeadline(httpContext, serverCallContext, handleCallTask);
+                }
+            }
+            catch (Exception ex)
+            {
+                serverCallContext.ProcessHandlerError(ex, Method.Name);
+                return Task.CompletedTask;
+            }
+
+            static async Task AwaitHandleCall(HttpContextServerCallContext serverCallContext, Method<TRequest, TResponse> method, Task handleCall)
+            {
+                try
+                {
+                    await handleCall;
+                    await serverCallContext.EndCallAsync();
+                }
+                catch (Exception ex)
+                {
+                    serverCallContext.ProcessHandlerError(ex, method.Name);
+                }
+            }
+        }
+
+        private async Task HandleCallWithDeadline(HttpContext httpContext, HttpContextServerCallContext serverCallContext, Task handleCallTask)
+        {
+            // CTS is used to...
+            // 1. Cancel Task.Delay if the handler completes first
+            // 2. CT is set on ServerCallContext to indicate deadline/request abort
+            var cts = new CancellationTokenSource();
+            CancellationTokenRegistration registration = default;
+
+            serverCallContext.SetCancellationToken(cts.Token);
+
+            try
+            {
+                // If the request is aborted then cancel the CTS
+                if (httpContext.RequestAborted.CanBeCanceled)
+                {
+                    registration = httpContext.RequestAborted.Register((c) =>
+                    {
+                        ((CancellationTokenSource)c).Cancel();
+                    }, cts, false);
+                }
+
+                var completedTask = await Task.WhenAny(
+                    handleCallTask,
+                    Task.Delay(serverCallContext.Timeout, cts.Token));
+
+                // Either the call handler is complete, and we want to cancel the Task.Delay
+                // Or the deadline has been exceeded and we want to trigger ServerCallContext.CancellationToken
+                cts.Cancel();
+
+                if (completedTask != handleCallTask)
+                {
+                    serverCallContext.DeadlineExceeded();
+                    httpContext.Response.ConsolidateTrailers(serverCallContext);
+
+                    // Ensure any errors thrown by the dead call are handled
+                    _ = ObserveDeadCall(serverCallContext, handleCallTask);
```

Comment:
This looks like a bad time. What's the expectation when user code hangs and you tear everything down? Just let it hang...

Reply:
Is it worse or better than HttpContext.Abort? Is there a difference between:
1. HttpContext.Abort being called, the server returning a response, and the request delegate hanging
2. The deadline being hit, the request delegate sending deadline response and exiting, and the call logic hanging in its own task

Current behavior:

> In gRPC there is the concept of a deadline. The deadline is sent by the client to the server with a gRPC call in a header. It is the client saying to the server "you have X seconds to finish this call, after-which give up".
>
> * If the server finished the call in less than X sends then the response completes as usual.
> * If the server exceeds the X seconds then it immediately finishes the HTTP/2 response, sending a RST_STREAM. The user's gRPC method code will likely still be executing after this. If they are good then it will be using a cancellation token, and they stop whatever they're doing quickly. But they might not, and continue doing stuff after the response is complete. This is fine. It is also fine if an error eventually gets thrown because they tried to do an HTTP thing.
>
> Today the gRPC server is finishing the connection when the deadline is hit by calling `HttpContext.Abort()`. This immediately sends RST_STREAM with an INTERNAL_ERROR error code. The request delegate with the user's code is still running. Using `Abort` is not ideal, but its response seems to be acceptable to the various gRPC clients of the world.

Reply:
I don't think swallowing and logging the errors is enough considering our plans for more HTTP/2 pooling in the future. If we ever want to support pooling HTTP/2 streams without the extra allocations required for poisoning the old streams, we'll want to ensure any calls that could be referencing the HttpContext complete before the middleware does. We don't want our servers to think they can pool the HttpContext until all calls associated with the request complete.